README.md
# µTask, the Lightweight Automation Engine

[![Build Status](https://travis-ci.org/ovh/utask.svg?branch=master)](https://travis-ci.org/ovh/utask)
[![Go Report Card](https://goreportcard.com/badge/github.com/ovh/utask)](https://goreportcard.com/report/github.com/ovh/utask)
[![Coverage Status](https://coveralls.io/repos/github/ovh/utask/badge.svg?branch=master)](https://coveralls.io/github/ovh/utask?branch=master)
[![GoDoc](https://godoc.org/github.com/ovh/utask?status.svg)](https://godoc.org/github.com/ovh/utask)
[![GitHub stars](https://img.shields.io/github/stars/ovh/utask)](https://github.com/ovh/utask/stargazers)
![GitHub last commit](https://img.shields.io/github/last-commit/ovh/utask)
[![GitHub license](https://img.shields.io/github/license/ovh/utask)](https://github.com/ovh/utask/blob/master/LICENSE)

µTask is an automation engine built for the cloud. It is:
- **simple to operate**: only a postgres DB is required
- **secure**: all data is encrypted, only visible to authorized users
- **extensible**: you can develop custom actions in golang

µTask allows you to model business processes in a **declarative yaml format**. Describe a set of inputs and a graph of actions and their inter-dependencies: µTask will asynchronously handle the execution of each action, working its way around transient errors and keeping an encrypted, auditable trace of all intermediary states until completion.

<img src="./assets/img/utask.png" width="50%" align="right">

## Table of contents

- [Real-world examples](#examples)
- [Quick Start](#quickstart)
- [Operating in production](#operating)
- [Configuration](#configuration)
- [Authoring Task Templates](#templates)
- [Extending µTask with plugins](#plugins)
- [Contributing](#contributing)
- [License](#license)

## Real-world examples <a name="examples"></a>

Here are a few real-world examples that can be implemented with µTask:

### Kubernetes ingress TLS certificate provisioning

A new ingress is created on the production kubernetes cluster. A hook triggers a µTask template that:
- generates a private key
- requests a new certificate
- meets the certificate issuer's challenges
- commits the resulting certificate back to the cluster

### New team member bootstrap

A new member joins the team. The team leader starts a task specifying the new member's name, that:
- asks the new team member to generate an SSH key pair and copy the public key in a µTask-generated form
- registers the public SSH key centrally
- creates accounts on internal services (code repository, CI/CD, internal PaaS, ...) for the new team member
- triggers another task to spawn a development VM
- sends a welcome email full of GIFs

### Payments API asynchronous processing

The payments API receives a request that requires an asynchronous antifraud check. It spawns a task on its companion µTask instance that:
- calls a first risk-assessing API which returns a number
- if the risk is low, the task succeeds immediately
- otherwise it calls a SaaS antifraud solution API which returns a score
- if the score is good, the task succeeds
- if the score is very bad, the task fails
- if it is in between, it triggers a human investigation step where an operator can enter a score in a µTask-generated form
- when it is done, the task sends an event to the payments API to notify of the result

The payments API keeps a reference to the running workflow via its task ID. Operators of the payments API can follow the state of current tasks by requesting the µTask instance directly. Depending on the payments API implementation, it may allow its callers to follow a task's state.

## Quick start <a name="quickstart"></a>

### Running with docker-compose

Download our latest install script, setup your environment and launch your own local instance of µTask.

```bash
mkdir utask && cd utask
wget https://github.com/ovh/utask/releases/latest/download/install-utask.sh
sh install-utask.sh
docker-compose up
```

All the configuration for the application is found in the environment variables in docker-compose.yaml. You'll see that basic auth is setup for user `admin` with password `1234`. Try logging in with this user on the graphical dashboard: [http://localhost:8081/ui/dashboard](http://localhost:8081/ui/dashboard).

You can also explore the API schema: [http://localhost:8081/unsecured/spec.json](http://localhost:8081/unsecured/spec.json).

Request a new task:
![](./assets/img/utask_new_task.png)

Get an overview of all tasks:
![](./assets/img/utask_dashboard.png)

Get a detailed view of a running task:
![](./assets/img/utask_running.png)

Browse available task templates:
![](./assets/img/utask_templates.png)

### Running with your own postgres service

Alternatively, you can clone this repository and build the µTask binary:

```bash
make all
```

## Operating in production <a name="operating"></a>

The folder you created in the previous step is meant to become a git repo where you version your own task templates and plugins. Re-download and run the latest install script to bump your version of µTask.

You'll deploy your version of µTask by building a docker image based on the official µTask image, which will include your extensions. See the Dockerfile generated during installation.

### Architecture

µTask is designed to run a task scheduler and perform the task workloads within a single runtime: work is not delegated to external agents. Multiple instances of the application will coordinate around a single postgres database: each will be able to determine independently which tasks are available. When an instance of µTask decides to execute a task, it will take hold of that task to avoid collisions, then release it at the end of an execution cycle.

A task will keep running as long as its steps are successfully executed. If a task's execution is interrupted before completion, it will become available to be re-collected by one of the active instances of µTask. That means that execution might start in one instance and resume on a different one.

### Maintenance procedures

#### Key rotation

1. Generate a new key with [symmecrypt](https://github.com/ovh/symmecrypt), with the 'storage' label.
2. Add it to your configuration items. The library will take all keys into account and use the latest possible key, falling back to older keys when finding older data.
3. Set your API in maintenance mode (env var or command line arg, see config below): all write actions will be refused when you reboot the API.
4. Reboot API.
5. Make a POST request on the /key-rotate endpoint of the API.
6. All data will be encrypted with the latest key, you can delete older keys.
7. De-activate maintenance mode.
8. Reboot API.

## Configuration 🔨 <a name="configuration"></a>

### Command line args

The µTask binary accepts the following arguments as binary args or env var.
All are optional and have a default value:
- `init-path`: the directory from where initialization plugins (see "Developing plugins") are loaded in *.so form (default: `./init`)
- `plugins-path`: the directory from where action plugins (see "Developing plugins") are loaded in *.so form (default: `./plugins`)
- `templates-path`: the directory where yaml-formatted task templates are loaded from (default: `./templates`)
- `functions-path`: the directory where yaml-formatted functions templates are loaded from (default: `./functions`)
- `region`: an arbitrary identifier, to aggregate a running group of µTask instances (commonly containers), and differentiate them from another group, in a separate region (default: `default`)
- `http-port`: the port on which the HTTP API listents (default: `8081`)
- `debug`: a boolean flag to activate verbose logs (default: `false`)
- `maintenance-mode`: a boolean to switch API to maintenance mode (default: `false`)

### Config keys and files

Checkout the [µTask config keys and files README](./config/README.md).

### Authentication

The vanilla version of µTask doesn't handle authentication by itself, it is meant to be placed behind a reverse proxy that provides a username through the "x-remote-user" http header. A username found there will be trusted as is, and used for authorization purposes (admin actions, task resolution, etc...).

For development purposes, an optional `basic-auth` configstore item can be provided to define a mapping of usernames and passwords. This is not meant for use in production.

Extending this basic authentication mechanism is possible by developing an "init" plugin, as described [below](#plugins).

### Notification

Every task state change can be notified to a notification backend.
µTask implements three differents notification backends: Slack, [TaT](https://github.com/ovh/tat), and generic webhooks.

Default payload that will be sent for generic webhooks are:

__task_state_update notifications:__
```json
{
    "message": "string",
    "notification_type": "task_state_update",
    "task_id": "public_task_uuid",
    "title": "task title string",
    "state": "current task state",
    "template": "template_name",
    "requester": "optional",
    "resolver": "optional",
    "steps": "14/20",
    "potential_resolvers": "user1,user2,admin",
    "resolution_id": "optional,public_resolution_uuid",
    "tags": "{\"tag1\":\"value1\"}"
}
```
__task_validation notifications:__
```json
{
    "message": "string",
    "notification_type": "task_validation",
    "task_id": "public_task_uuid",
    "title": "task title string",
    "state": "TODO",
    "template": "template_name",
    "requester": "optional",
    "potential_resolvers": "user1,user2,admin",
    "tags": "{\"tag1\":\"value1\"}"
}
```

Notification backends can be configured in the global µTask configuration, as described [here](./config/README.md#utask-cfg).

## Authoring Task Templates <a name="templates"></a>

Checkout the [µTask examples directory](./examples).

A process that can be executed by µTask is modelled as a `task template`: it is written in yaml format and describes a sequence of steps, their interdepencies, and additional conditions and constraints to control the flow of execution.

The user that creates a task is called `requester`, and the user that executes it is called `resolver`. Both can be the same user in some scenarios.

A user can be allowed to resolve a task in three ways:
- the user is included in the global configuration's list of `admin_usernames`
- the user is included in the task's template list of `allowed_resolver_usernames`
- the user is included in the task `resolver_usernames` list

### Value Templating

µTask uses the go [templating engine](https://golang.org/pkg/text/template/) in order to introduce dynamic values during a task's execution. As you'll see in the example template below, template handles can be used to access values from different sources. Here's a summary of how you can access values through template handles:
- `.input.[INPUT_NAME]`: the value of an input provided by the task's requester
- `.resolver_input.[INPUT_NAME]`: the value of an input provided by the task's resolver
- `.step.[STEP_NAME].output.foo`: field `foo` from the output of a named step
- `.step.[STEP_NAME].metadata.HTTPStatus`: field `HTTPStatus` from the metadata of a named step
- `.step.[STEP_NAME].children`: the collection of results from a 'foreach' step
- `.step.[STEP_NAME].error`: error message from a failed step
- `.step.[STEP_NAME].state`: current state of the given step
- `.config.[CONFIG_ITEM].bar`: field `bar` from a config item (configstore, see above)
- `.iterator.foo`: field `foo` from the iterator in a loop (see `foreach` steps below)
- `.pre_hook.output.foo`: field `foo` from the output of the step's preHook (see `preHooks` below)
- `.pre_hook.metadata.HTTPStatus`: field `HTTPStatus` from the metadata of the step's preHook (see `preHooks` below)
- `.function_args.[ARG_NAME]`: argument that needs to be given in the conifguration section to the function (see `functions` below)

The following templating functions are available:

| Name               | Description                                                                                                                                                                                                                                                                                                                                                     | Reference                                                |
| ------------------ | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------- |
| **`Golang`**       | Builtin functions from Golang text template                                                                                                                                                                                                                                                                                                                     | [Doc](https://golang.org/pkg/text/template/#hdr-Actions) |
| **`Sprig`**        | Extended set of functions from the Sprig project                                                                                                                                                                                                                                                                                                                | [Doc](https://masterminds.github.io/sprig/)              |
| **`field`**        | Equivalent to the dot notation, for entries with forbidden characters                                                                                                                                                                                                                                                                                           | ``{{field `config` `foo.bar`}}``                         |
| **`fieldFrom`**    | Equivalent to the dot notation, for entries with forbidden characters. It takes the previous template expression as source for the templating values. Example: ```{{ `{"foo.foo":"bar"}` | fromJson | fieldFrom `foo.foo` }}```                                                                                                                                   | ```{{expr | fieldFrom `config` `foo.bar`}}```              |
| **`eval`**         | Evaluates the value of a template variable                                                                                                                                                                                                                                                                                                                      | ``{{eval `var1`}}``                                      |
| **`evalCache`**    | Evaluates the value of a template variable, and cache for future usage (to avoid further computation)                                                                                                                                                                                                                                                           | ``{{evalCache `var1`}}``                                 |
| **`fromJson`**     | Decodes a JSON document into a structure. If the input cannot be decoded as JSON, the function will return an empty string                                                                                                                                                                                                                                      | ``{{fromJson `{"a":"b"}`}}``                             |

CHANGELOG.md
# CHANGELOG

A detailed changelog is available in our [GitHub release page](https://github.com/ovh/utask/releases "GitHub release page").

We will list in this document any breaking changes between versions, that require modifications of the task templates you maintain.

## Breaking changes

### v1.13.0
#### Notifications
- Added a new `notification_type` : `task_validation` that fires every time a new task need a human validation. To integrate different `notification_strategy`, all `notification_strategy` are scopped to the `notification_type`. Then, `default_notification_strategy` is now an object containing the `notification_type` as key, and the `strategy` as value ; and `template_notification_strategies` is now an object containing the `notification_type` as key, and the strategies array as value.
- Changed the configuration key `task_state_action` (in the `notify_action` section) to the new value `task_state_update` (to match the `notification_type` value).

### v1.10.0
#### SQL
- `005_resolution_creation_timestamp.sql` migration file should be applied while upgrading. It adds a column `created` in the `resolution` table to keep track of the resolution creation timestamp.

#### Prometheus metrics
- The Prometheus gauges used to record the number of tasks by state haven been moved to a single `GaugeVec` named `utask_task_state`.

### v1.5.0 (2020-04-09)
##### Templating
- `jsonmarshal` function has been __deleted__. It was obsolete since the `sprig` package used by utask provides many JSON related functions (e.g. `toJson`).
- `jsonfield` function has been __deleted__. It was obsolete since a pipeline such as ``{{ field `foo` | toJson }}`` achieves the same thing.

##### API model
- ``/resolution``: `resolution.payload` (deprecated, duplicate of / replaced by `resolution.output` since 1.0) has finally been __deleted__.

### v1.4.0 (2020-03-31)
##### API routes
- `GET /resolution` API route has been __deleted__. We decided that resolutions should always be accessed through tasks

##### Configuration
- `resolver_usernames` configuration field __deleted__. Global resolver is not a feature we want to support, and its usage was misguided.

##### SQL
- `001_resolver_watcher_usernames_indexes.sql` migration file should be applied while upgrading. This migration is non-breaking but will boost performance on big instances for tasks listing.

### v1.3.0 (2020-03-17)
##### HTTP plugin
- `timeout_seconds` configuration field __deleted__. It has been replaced by the field `timeout`, using the Golang time.Duration format (5s, 1m, ...), for consistency with other plugins and to express timeout durations inferior than 1 second. #86
- `parameters` configuration field __deleted__. It has been replaced by the field `query_parameters`: the name is now clearer, and the object format is similar to the `headers` configuration field. #86
- `deny_redirects` configuration field __deleted__. It has been replaced by the field `follow_redirect` : the new default behavior will be to never follow redirections, unless specified by this configuration field. #86

##### SSH plugin
- `allow_exit_non_zero` configuration field __deleted__. This field has no strict replacement, as its behavior was not in the uTask philosophy. The field `exit_codes_unrecoverable` has been introduced to catch some exit codes as `CLIENT_ERROR` if the error is not recoverable (to either halt the execution or change to a custom status via check conditions). #85
- `exit_status` metadata field __renamed__ `exit_code`. This field's name is now consistent between the ssh and script plugins. #85

##### script plugin
- `allow_exit_non_zero` configuration field __deleted__. This field has no strict replacement, as its behavior was not in the uTask philosophy. The field `exit_codes_unrecoverable` has been introduced to catch some exit codes as `CLIENT_ERROR` if the error is not recoverable (to either halt the execution or change to a custom status via check conditions). #87
- `last_line_not_json` configuration field __deleted__. It has been replaced by the field `output_mode` which supports more options, and can be configured to the value `disabled` to reproduce the same behavior. #87


api/server.go
package api

import (
	"context"
	"fmt"
	"net/http"
	"os"
	"os/signal"
	"path"
	"strings"
	"syscall"

	"github.com/gin-gonic/gin"
	"github.com/loopfz/gadgeto/tonic"
	"github.com/loopfz/gadgeto/tonic/utils/jujerr"
	"github.com/loopfz/gadgeto/zesty"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"github.com/sirupsen/logrus"
	"github.com/wI2L/fizz"
	"github.com/wI2L/fizz/openapi"

	"github.com/ovh/utask"
	"github.com/ovh/utask/api/handler"
	"github.com/ovh/utask/models/resolution"
	"github.com/ovh/utask/models/task"
	"github.com/ovh/utask/pkg/auth"
)

// Server wraps the http handler that exposes a REST API to control
// the task orchestration engine
type Server struct {
	httpHandler            *fizz.Fizz
	authMiddleware         func(*gin.Context)
	dashboardPathPrefix    string
	dashboardAPIPathPrefix string
	dashboardSentryDSN     string
	editorPathPrefix       string
	maxBodyBytes           int64
	customMiddlewares      []gin.HandlerFunc
}

// NewServer returns a new Server
func NewServer() *Server {
	return &Server{
		authMiddleware: func(c *gin.Context) { c.Next() }, // default no-op middleware
	}
}

// WithAuth configures the Server's auth middleware
// it receives an authProvider function capable of extracting a caller's identity from an *http.Request
// the authProvider function also has discretion to deny authorization for a request by returning an error
func (s *Server) WithAuth(authProvider func(*http.Request) (string, error)) {
	if authProvider != nil {
		s.authMiddleware = authMiddleware(authProvider)
	}
}

// WithCustomMiddlewares sets an array of customized gin middlewares.
// It helps for init plugins to include these customized middlewares in the api server
func (s *Server) WithCustomMiddlewares(customMiddlewares ...gin.HandlerFunc) {
	s.customMiddlewares = append(s.customMiddlewares, customMiddlewares...)
}

// SetDashboardPathPrefix configures the custom path prefix for dashboard static files hosting.
// It doesn't change the path used by utask API to serve the files, it's only used inside UI files
// in order that dashboard can be aware of a ProxyPass configuration.
func (s *Server) SetDashboardPathPrefix(dashboardPathPrefix string) {
	s.dashboardPathPrefix = dashboardPathPrefix
}

// SetDashboardAPIPathPrefix configures a custom path prefix that UI should use when calling utask API.
// Required when utask API is exposed behind a ProxyPass and UI need to know the absolute URI to call.
func (s *Server) SetDashboardAPIPathPrefix(dashboardAPIPathPrefix string) {
	s.dashboardAPIPathPrefix = dashboardAPIPathPrefix
}

// SetDashboardSentryDSN configures a Sentry DSN URI to send UI exceptions and failures to.
func (s *Server) SetDashboardSentryDSN(dashboardSentryDSN string) {
	s.dashboardSentryDSN = dashboardSentryDSN
}

// SetEditorPathPrefix configures a custom path prefix for editor static files hosting.
// It doesn't change the path used by utask API to serve the files, it's only used inside UI files
// in order that editor can be aware of a ProxyPass configuration.
func (s *Server) SetEditorPathPrefix(editorPathPrefix string) {
	s.editorPathPrefix = editorPathPrefix
}

// SetMaxBodyBytes
func (s *Server) SetMaxBodyBytes(max int64) {
	s.maxBodyBytes = max
}

// ListenAndServe launches an http server and stays blocked until
// the server is shut down by a system signal
func (s *Server) ListenAndServe() error {
	ctx, cancel := context.WithCancel(context.Background())

	s.build(ctx)
	stop := make(chan os.Signal, 1)
	signal.Notify(stop, os.Interrupt, syscall.SIGTERM)
	srv := &http.Server{Addr: fmt.Sprintf(":%d", utask.FPort), Handler: s.httpHandler}

	go func() {
		<-stop
		logrus.Info("Shutting down...")
		cancel()

		if err := srv.Shutdown(context.Background()); err != nil {
			logrus.Fatal(err)
		}
	}()

	if err := srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
		return err
	}
	return nil
}

// Handler returns the underlying http.Handler of a Server
func (s *Server) Handler(ctx context.Context) http.Handler {
	s.build(ctx)
	return s.httpHandler
}

func generateBaseHref(pathPrefix, uri string) string {
	// UI requires to have a trailing slash at the end
	return path.Join(pathPrefix, uri) + "/"
}

func generatePathPrefixAPI(pathPrefix string) string {
	p := path.Join(pathPrefix, "/")
	if p == "." {
		p = "/"
	} else if !strings.HasSuffix(p, "/") {
		p += "/"
	}
	return p
}

// build registers all routes and their corresponding handlers for the Server's API
func (s *Server) build(ctx context.Context) {
	if s.httpHandler == nil {
		ginEngine := gin.Default()

		ginEngine.
			Group("/",
				StaticFilePatternReplaceMiddleware(
					"___UTASK_DASHBOARD_BASEHREF___",
					generateBaseHref(s.dashboardPathPrefix, "/ui/dashboard"),
					"___UTASK_DASHBOARD_PREFIXAPIBASEURL___",
					generatePathPrefixAPI(s.dashboardAPIPathPrefix),
					"___UTASK_DASHBOARD_SENTRY_DSN___",
					s.dashboardSentryDSN)).
			StaticFS("/ui/dashboard", http.Dir("./static/dashboard"))

		ginEngine.Group("/",
			StaticFilePatternReplaceMiddleware(
				"___UTASK_EDITOR_BASEHREF___",
				generateBaseHref(s.editorPathPrefix, "/ui/editor"))).
			StaticFS("/ui/editor", http.Dir("./static/editor"))

		ginEngine.
			Group("/",
				StaticFilePatternReplaceMiddleware(
					"___UTASK_DASHBOARD_PREFIXAPIBASEURL___",
					generatePathPrefixAPI(s.dashboardAPIPathPrefix),
					"___UTASK_DASHBOARD_SENTRY_DSN___",
					s.dashboardSentryDSN)).
			StaticFS("/ui/swagger", http.Dir("./static/swagger-ui"))

		collectMetrics(ctx)
		ginEngine.GET("/metrics", gin.WrapH(promhttp.Handler()))

		router := fizz.NewFromEngine(ginEngine)

		openapiPathPrefix := s.dashboardAPIPathPrefix
		if openapiPathPrefix == "" {
			openapiPathPrefix = "/"
		}
		router.Generator().SetServers([]*openapi.Server{
			{
				URL:         openapiPathPrefix,
				Description: utask.AppName(),
			},
		})

		router.Use(s.customMiddlewares...)
		router.Use(ajaxHeadersMiddleware, errorLogMiddleware)

		tonic.SetErrorHook(jujerr.ErrHook)
		tonic.SetBindHook(yamlBindHook(s.maxBodyBytes))
		tonic.SetRenderHook(yamljsonRenderHook, "application/json")

		authRoutes := router.Group("/", "x-misc", "Misc authenticated routes", s.authMiddleware)
		{
			templateRoutes := authRoutes.Group("/", "04 - template", "Manage uTask task templates")
			{
				// public template listing
				templateRoutes.GET("/template",
					[]fizz.OperationOption{
						fizz.Summary("List task templates"),
					},
					tonic.Handler(handler.ListTemplates, 200))
				templateRoutes.GET("/template/:name",
					[]fizz.OperationOption{
						fizz.Summary("Get task template details"),
					},
					tonic.Handler(handler.GetTemplate, 200))
			}

			functionRoutes := authRoutes.Group("/", "05 - function", "Manage uTask task functions")
			{
				// public function listing
				functionRoutes.GET("/function",
					[]fizz.OperationOption{
						fizz.Summary("List task functions"),
					},
					tonic.Handler(handler.ListFunctions, 200))
				functionRoutes.GET("/function/:name",
					[]fizz.OperationOption{
						fizz.Summary("Get task function details"),
					},
					tonic.Handler(handler.GetFunction, 200))
			}

			// task
			taskRoutes := authRoutes.Group("/", "01 - task", "Manage uTask tasks")
			{
				// task creation in batches
				taskRoutes.POST("/batch",
					[]fizz.OperationOption{
						fizz.Summary("Create a batch of tasks"),
					},
					maintenanceMode,
					tonic.Handler(handler.CreateBatch, 201))
				taskRoutes.POST("/task",
					[]fizz.OperationOption{
						fizz.Summary("Create new task"),
					},
					maintenanceMode,
					tonic.Handler(handler.CreateTask, 201))
				taskRoutes.GET("/task",
					[]fizz.OperationOption{
						fizz.Summary("List tasks"),
					},
					tonic.Handler(handler.ListTasks, 200))
				taskRoutes.GET("/task/:id",
					[]fizz.OperationOption{
						fizz.Summary("Get task details"),
					},
					tonic.Handler(handler.GetTask, 200))
				taskRoutes.PUT("/task/:id",
					[]fizz.OperationOption{
						fizz.Summary("Edit task"),
					},
					maintenanceMode,
					tonic.Handler(handler.UpdateTask, 200))
				taskRoutes.POST("/task/:id/wontfix",
					[]fizz.OperationOption{
						fizz.Summary("Cancel task"),
					},
					maintenanceMode,
					tonic.Handler(handler.WontfixTask, 204))
				taskRoutes.DELETE("/task/:id",
					[]fizz.OperationOption{
						fizz.Summary("Delete task"),
						fizz.Description("Admin rights required"),
					},
					requireAdmin,
					maintenanceMode,
					tonic.Handler(handler.DeleteTask, 204))
			}

			// comments
			commentsRoutes := authRoutes.Group("/", "03 - comment", "Manage uTask task comments")
			{
				commentsRoutes.POST("/task/:id/comment",
					[]fizz.OperationOption{
						fizz.Summary("Post new comment on task"),
					},
					maintenanceMode,
					tonic.Handler(handler.CreateComment, 201))
				commentsRoutes.GET("/task/:id/comment",
					[]fizz.OperationOption{
						fizz.Summary("List task comments"),
					},
					tonic.Handler(handler.ListComments, 200))
				commentsRoutes.GET("/task/:id/comment/:commentid",
					[]fizz.OperationOption{
						fizz.Summary("Get single task comment"),
					},
					tonic.Handler(handler.GetComment, 200))
				commentsRoutes.PUT("/task/:id/comment/:commentid",
					[]fizz.OperationOption{
						fizz.Summary("Edit task comment"),
					},
					maintenanceMode,
					tonic.Handler(handler.UpdateComment, 200))
				commentsRoutes.DELETE("/task/:id/comment/:commentid",
					[]fizz.OperationOption{
						fizz.Summary("Delete task comment"),
					},
					maintenanceMode,
					tonic.Handler(handler.DeleteComment, 204))
			}

			// resolution
			resolutionRoutes := authRoutes.Group("/", "02 - resolution", "Manager uTask resolutions")
			{
				resolutionRoutes.POST("/resolution",
					[]fizz.OperationOption{
						fizz.Summary("Create task resolution"),
						fizz.Summary("This action instantiates a holder for the task's execution state. Only an approved resolver or admin user can perform this action."),
					},
					maintenanceMode,
					tonic.Handler(handler.CreateResolution, 201))
				resolutionRoutes.GET("/resolution/:id",
					[]fizz.OperationOption{
						fizz.Summary("Get the details of a task resolution"),
						fizz.Description("Details include the intermediate results of every step. Admin users can view any resolution's details."),
					},
					tonic.Handler(handler.GetResolution, 200))
				resolutionRoutes.PUT("/resolution/:id",
					[]fizz.OperationOption{
						fizz.Summary("Edit a task's resolution during execution."),
						fizz.Description("Action of last resort if a task needs fixing. Admin users only."),
					},
					requireAdmin,
					maintenanceMode,
					tonic.Handler(handler.UpdateResolution, 204))
				resolutionRoutes.POST("/resolution/:id/run",
					[]fizz.OperationOption{
						fizz.Summary("Execute a task"),
					},
					maintenanceMode,
					tonic.Handler(handler.RunResolution, 204))
				resolutionRoutes.POST("/resolution/:id/pause",
					[]fizz.OperationOption{
						fizz.Summary("Pause a task's execution"),
						fizz.Description("This action takes a task out of the execution pipeline, it will not be considered for automatic retry until it is re-run manually."),
					},
					maintenanceMode,
					tonic.Handler(handler.PauseResolution, 204))
				resolutionRoutes.POST("/resolution/:id/extend",
					[]fizz.OperationOption{
						fizz.Summary("Extend max retry limit for a task's execution"),
					},
					maintenanceMode,
					tonic.Handler(handler.ExtendResolution, 204))
				resolutionRoutes.POST("/resolution/:id/cancel",
					[]fizz.OperationOption{
						fizz.Summary("Cancel a task's execution"),
					},
					maintenanceMode,
					tonic.Handler(handler.CancelResolution, 204))
				resolutionRoutes.GET("/resolution/:id/step/:stepName",
					[]fizz.OperationOption{
						fizz.Summary("Get the details of the step of a task resolution"),
						fizz.Description("Returns the current implementation of the step, including the output of the step."),
					},
					tonic.Handler(handler.GetResolutionStep, 200))
				resolutionRoutes.PUT("/resolution/:id/step/:stepName",
					[]fizz.OperationOption{
						fizz.Summary("Edit the details of the step of a task resolution"),
						fizz.Description("Allow the edition of a step, if a step needs fixing. Admin users only."),
					},
					requireAdmin,
					maintenanceMode,
					tonic.Handler(handler.UpdateResolutionStep, 204))
				resolutionRoutes.PUT("/resolution/:id/step/:stepName/state",
					[]fizz.OperationOption{
						fizz.Summary("Edit the state of the step of a task resolution"),
						fizz.Description("Allow the edition of the step state, if a step needs to be re-run or skipped manually. Resolution managers only."),
					},
					maintenanceMode,
					tonic.Handler(handler.UpdateResolutionStepState, 204))

				//	resolutionRoutes.POST("/resolution/:id/rollback",
				//		[]fizz.OperationOption{
				// 			fizz.Summary(""),
				//		},
				//		tonic.Handler(handler.ResolutionRollback, 200))
			}

			authRoutes.GET("/",
				[]fizz.OperationOption{
					fizz.Summary("Redirect to /meta"),
				},
				func(c *gin.Context) {
					c.Redirect(http.StatusMovedPermanently, "/meta")
				})

			authRoutes.GET("/meta",
				[]fizz.OperationOption{
					fizz.Summary("Display service name and user's status"),
				},
				tonic.Handler(rootHandler, 200))

			// admin
			authRoutes.POST("/key-rotate",
				[]fizz.OperationOption{
					fizz.Summary("Re-encrypt all data with latest storage key"),
				},
				requireAdmin,
				tonic.Handler(keyRotate, 200))
		}

		router.GET("/unsecured/mon/ping",
			[]fizz.OperationOption{
				fizz.Summary("Assert that the service is running and can talk to it's data backend"),
			},
			pingHandler)
		router.GET("/unsecured/spec.json", nil, router.OpenAPI(&openapi.Info{
			Title:   utask.AppName(),
			Version: utask.Version,
		}, "json"))
		router.GET("/unsecured/stats",
			[]fizz.OperationOption{
				fizz.Summary("Fetch statistics about existing tasks"),
			},
			tonic.Handler(Stats, 200))

		s.httpHandler = router
	}
}

func pingHandler(c *gin.Context) {
	dbp, err := zesty.NewDBProvider(utask.DBName)
	if err != nil {
		c.String(http.StatusInternalServerError, "")
		c.Error(err)
		return
	}
	i, err := dbp.DB().SelectInt(`SELECT 1`)
	if err != nil {
		c.String(http.StatusInternalServerError, "")
		c.Error(err)
		return
	}
	if i != 1 {
		c.String(http.StatusInternalServerError, "")
		c.Error(fmt.Errorf("Unexpected value %d", i))
		return
	}
	c.String(http.StatusOK, "pong")
}

type rootOut struct {
	ApplicationName string `json:"application_name"`
	UserIsAdmin     bool   `json:"user_is_admin"`
	Username        string `json:"username"`
	Version         string `json:"version"`
	Commit          string `json:"commit"`
}

func rootHandler(c *gin.Context) (*rootOut, error) {
	return &rootOut{
		ApplicationName: utask.AppName(),
		UserIsAdmin:     auth.IsAdmin(c) == nil,
		Username:        auth.GetIdentity(c),
		Version:         utask.Version,
		Commit:          utask.Commit,
	}, nil
}

func requireAdmin(c *gin.Context) {
	if err := auth.IsAdmin(c); err != nil {
		c.AbortWithStatus(http.StatusUnauthorized)
		return
	}
	c.Next()
}

func maintenanceMode(c *gin.Context) {
	if utask.FMaintenanceMode {
		c.JSON(http.StatusMethodNotAllowed, map[string]string{
			"error": "Maintenance mode activated",
		})
		return
	}
	c.Next()
}

func keyRotate(c *gin.Context) error {
	dbp, err := zesty.NewDBProvider(utask.DBName)
	if err != nil {
		return err
	}
	if err := task.RotateTasks(dbp); err != nil {
		return err
	}
	return resolution.RotateResolutions(dbp)
}
